feat: Add web extension support for syntax highlighting

mm-broadcom opened the pull request:
Adds support to build a web extension that contains a subset of features from the main Cobol LSP extension. Currently only includes the syntax highlighting.

## How Has This Been Tested?

This has been tested through building the extension and checking the functionality thereof within VS Code for Web following [this guide](https://code.visualstudio.com/api/extension-guides/web-extensions#test-your-web-extension). No unit tests were created this only provides the syntaxes developed for the main LSP extension within a web compatible extension.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
   - Not applicable. See the "How Has This Been Tested?" section above.
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Add web extension support for syntax highlighting

Adds support to build a web extension that contains a subset of features from the main Cobol LSP extension. Currently only includes the syntax highlighting.

VitGottwald commented:
Looks good. Do you think you could use `esbuild` for the bundling instead of `webpack`?

VitGottwald commented on clients/cobol-lsp-vscode-extension/src/web/test/suite/extension.test.ts:
```diff
@@ -0,0 +1,15 @@
+// import * as myExtension from '../../extension';
+
+suite('Web Extension Test Suite', () => {
+	vscode.window.showInformationMessage('Start all tests.');
```
Where and how is this test run in the CI?

mm-broadcom replied:
The tests for a web extension require it to be run in a web browser, according to [this documentation](https://code.visualstudio.com/api/extension-guides/web-extensions#web-extension-tests) from Microsoft.

Currently the unit test in the commit is only the one that was included in the template generation code given by Microsoft.

slavek-kucera replied:
For testing in CI, https://github.com/microsoft/vscode-test-web package can be used.

VitGottwald commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -37,6 +37,7 @@
   ],
   "icon": "resources/logo.png",
   "main": "./dist/extension.js",
+  "browser": "./dist/web/extension.js",
```
How is this bundled in the CI?

slavek-kucera replied:
We'll need an extra step in `.github/workflows/build.yml` that will produce a new vsix package, packaged with `--target web` option.

VitGottwald replied:
Let's do that. We will need it for publishing the WEB vsix.

mm-broadcom replied:
Added the step to run the package for web.

mm-broadcom committed:
Limit commands in web context

Limit/Disable the commands available when it is run as a web extension. Add documentation denoting that the extension's capabilities are limited as well.

VitGottwald commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -530,6 +547,7 @@
     "vscode:prepublish": "npm run -S build -- --minify",
     "postinstall": "node -e \"if (process.env.NODE_ENV === 'production'){process.exit(1)} \" && patch-package || echo skip patching",
     "build": "esbuild src/extension.ts --bundle --sourcemap --platform=node --target=node16.17.1 --outdir=dist --external:vscode --format=cjs",
+    "build:web": "esbuild src/web/extension.ts --bundle --sourcemap --platform=node --target=node16.17.1 --outdir=dist --external:vscode --format=cjs",
```
Does this actually run in the browser with `---platform=node` and `--format=cjs`?

mm-broadcom replied:
Yes it does. Initially found the build command [here](https://www.roboleary.net/2024/02/16/vscode-ext-esbuild.html), but I have also tested it myself. Outside of the myriad of restrictions that the web version has, there is not too much difference in the core code between it and the installed version as VS Code (installed) uses Electron/Chromium.

mm-broadcom committed:
Update web extension output directory

Updated the web extension's output directory to use the dist/web/ folder instead of the root of the dist one.

mm-broadcom committed:
Update Build.sh to build web extension

Updated the build.sh to build the web extension.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -530,6 +547,7 @@
     "vscode:prepublish": "npm run -S build -- --minify",
     "postinstall": "node -e \"if (process.env.NODE_ENV === 'production'){process.exit(1)} \" && patch-package || echo skip patching",
     "build": "esbuild src/extension.ts --bundle --sourcemap --platform=node --target=node16.17.1 --outdir=dist --external:vscode --format=cjs",
+    "build:web": "esbuild src/web/extension.ts --bundle --sourcemap --platform=node --target=node16.17.1 --outdir=dist/web --external:vscode --format=cjs",
```
The platform should be `browser`.

mm-broadcom replied:
Fixed in later commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -37,56 +37,73 @@
         "category": "COBOL",
-        "title": "Fetch Copybook"
+        "title": "Fetch Copybook",
+        "when": "!virtualWorkspace"
```
There is a variable `isWeb` that more appropriately represents the condition.

mm-broadcom replied:
The "!virtualWorkspace" was taken from Microsoft's documentation: https://code.visualstudio.com/api/extension-guides/virtual-workspaces#disable-functionality-when-a-virtual-workspace-is-opened

slavek-kucera replied:
virtual workspace can be opened even in the desktop vscode.

mm-broadcom replied:
Changed to use "!isWeb" in later commit.

mm-broadcom committed:
Change Web build command to "browser" as platform

Update the npm build:web command to use the browser as the platform rather than node.

mm-broadcom committed:
Update packaage.json to change "when" clauses

Changed the workspace detection in package.json to use the "isWeb" variable rather than virtual workspaces. (Web is a subset of use cases when the filesystem is virtual.)

mm-broadcom committed:
Add package command for web extension

Add the package command to create the web compatible extension. Also exclude the extension.js and associated source map from being ignored in vscodeignore.

mm-broadcom committed:
Replace webpack with esbuild

Replaced webpack with esbuild. Fixed test:web command such that the tests can run.

mm-broadcom committed:
Add web build to build.yml

Add web build step to the build.yml

VitGottwald commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -569,7 +589,9 @@
     "nyc": "^15.1.0",
     "patch-package": "^6.4.7",
     "prettier": "2.8.7",
+    "process": "^0.11.10",
```
Do we still need this after webpack has been removed?

VitGottwald replied:
And similarly for all the other dependencies added in the PR. Please, check if they can be removed. It looks like some of them are a ripple effect of having had webpack.

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
Remove unused Webpack-related packages

Removed unused Webpack-related packages.

mm-broadcom committed:
Add package:web step to Github workflow

Add package:web step to Github workflow

mm-broadcom committed:
Update package-lock.json

Update package-lock.json

mm-broadcom committed:
Revert "Update package-lock.json"

This reverts commit 53df54205f6026b844f8e80fa9a1a8285a9ed2b7.

mm-broadcom committed:
Updated package-lock.json

Updated package-lock.json

slavek-kucera commented on .github/workflows/build.yml:
```diff
@@ -341,6 +346,11 @@ jobs:
         run: |
           npm run package  -- --target ${{ env.target }}
           cp *.vsix ../../.
+      - name: Package COBOL LS vsix for Web
+        working-directory: clients/cobol-lsp-vscode-extension
+        run: |
+          npm run package:web
+          cp *.vsix ../../.
```
This will run 6x (once in each native configuration) and possibly overwrite the "desktop" VSIX package.
I suggest creating a separate job (e.g. buildWeb) in this file that will just do
- checkout
- "update version for PR" step
- npm ci
- npm run test:web
- npm run package:web
- upload artifact to preserve the VSIX

mm-broadcom replied:
It makes a separate vsix file with the name `cobol-language-support-web-[version]` where `version` is the latest numbered version. (2.1.2 as of this writing.)

mm-broadcom replied:
Made a separate job for building the web extension. Will need some help with the upload as there isn't an existing one to handle the artifact.

slavek-kucera replied:
https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/2c0b20fa249da679d33ac4fd9e5279dc1fd5eab4/.github/workflows/build.yml#L344-L348

slavek-kucera commented on clients/cobol-lsp-vscode-extension/esbuild.js:
```diff
@@ -0,0 +1,100 @@
+            polyfill.NodeGlobalsPolyfillPlugin({
+                process: true,
+                buffer: true,
+            }),
```
do we actually need the node polyfill?

mm-broadcom replied:
Removed in latest commit.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/esbuild.js:
```diff
@@ -0,0 +1,100 @@
+const path = require('path');
+const polyfill = require('@esbuild-plugins/node-globals-polyfill');
+
+const production = process.argv.includes('--production');
```
I don't think that the argument is actually provided in the build scripts.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/esbuild.js:
```diff
@@ -0,0 +1,100 @@
+            global: 'globalThis',
+        },
+
+        plugins: [
```
assert is not available in the web environment.
```
[watch] build started
✘ [ERROR] Could not resolve "assert"
    src/web/test/suite/extension.test.ts:1:24:
[watch] build finished
Error: Build failed with 1 error:
src/web/test/suite/extension.test.ts:1:24: ERROR: Could not resolve "assert"
    at failureErrorWithLog (C:\Source\che-che4z-lsp-for-cobol\clients\cobol-lsp-vscode-extension\node_modules\esbuild\lib\main.js:1651:15)
    at C:\Source\che-che4z-lsp-for-cobol\clients\cobol-lsp-vscode-extension\node_modules\esbuild\lib\main.js:1059:25
    at C:\Source\che-che4z-lsp-for-cobol\clients\cobol-lsp-vscode-extension\node_modules\esbuild\lib\main.js:1527:9
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  errors: [Getter/Setter],
  warnings: [Getter/Setter]
}
```

mm-broadcom replied:
It is because the package was removed from the DevDependencies in this earlier commit: https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/pull/2323/commits/c20d3181b8b15e0210b415758d8b5e90d129fca5
Is able to build once added back in. Fixed in latest commit.

mm-broadcom committed:
Fix assert build issues

Fix the issue with "Assert" not being available as a package.

mm-broadcom committed:
Remove Node Polyfill

Remove Node Polyfill

mm-broadcom committed:
Make buildWeb a separate job step for CI

Make buildWeb a separate job step for CI

mm-broadcom committed:
Ran Prettier

Ran prettier to fix the linting issues.

mm-broadcom committed:
Add upload step to buildWeb job

Add upload step to buildWeb job

slavek-kucera commented on .github/workflows/build.yml:
```diff
@@ -157,6 +157,27 @@ jobs:
 
+  buildWeb:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Unit Test COBOL LS Web extension
```
```suggestion
    steps:
      - uses: actions/checkout@v4
      - name: Unit Test COBOL LS Web extension
```

mm-broadcom committed:
Update .github/workflows/build.yml

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

slavek-kucera commented:
Please upgrade various ignore files...
`.prettierignore` - add `.vscode-test-web`
`jest.config.js` - add `"<rootDir>/src/web","<rootDir>/.vscode-test","<rootDir>/.vscode-test-web",`
`src/test/suite/index.ts` - `exclude: ["**/test/**", ".vscode-test/**", ".vscode-test-web/**"],`
`tsconfig.json` - `"exclude": ["node_modules", "server", ".vscode-test", ".vscode-test-web"]`

mm-broadcom committed:
TDD fixes for web extension

Implemented requested changes to TDD for it to account for the web extension.

mm-broadcom committed:
Merge branch 'web-extension-support' of https://github.com/mm-broadcom/che-che4z-lsp-for-cobol into web-extension-support

slavek-kucera commented on clients/cobol-lsp-vscode-extension/esbuild.js:
```diff
@@ -0,0 +1,99 @@
+const esbuild = require("esbuild");
```
It looks like the `esbuild.js` name interferes with the esbuild package. I suggest renaming to e.g. `build.web.js`.

mm-broadcom replied:
Resolved in latest commit.

mm-broadcom committed:
Rename esbuild.js

Rename esbuild.js to avoid collision with just "esbuild".

mm-broadcom committed:
Update .vscodeignore

Update .vscodeignore to exclude extension related javascript and map files from the extension/dist/ folders.

slavek-kucera commented:
Look through the build logs, I think that the web version also builds the desktop version in the `vscode:prepublish` step.
I also think that the `browser` entrypoint in the package.json will have to be deleted for the non-browser build and the `main` entrypoint in the web version.

mm-broadcom committed:
Fix web build and package

Fix web build and package.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -538,7 +538,7 @@
   },
   "scripts": {
-    "vscode:prepublish": "npm run -S build -- --minify",
+    "vscode:prepublish": "npm run -S build -- --minify && npm run -S build:web -- --minify",
```
```suggestion
    "vscode:prepublish": "npm run -S build -- --minify && npm run -S build:web -- --production",
```
?

mm-broadcom replied:
Was only repeating the flags set for the regular build. Verified that the change worked locally as I was able to also replicate the issue. Arose from the packaging step not having the web build's files to go off of.

mm-broadcom replied:
Added the flag for production but also kept the minify flag.

mm-broadcom committed:
Add production flag to web build in prepackage

Add production flag to web build in prepackage

slavek-kucera commented on clients/cobol-lsp-vscode-extension/.vscodeignore:
```diff
@@ -37,6 +37,12 @@
 !dist/extension.js.map
+!dist/web/extension.js
+!dist/web/extension.js.map
+!extension/dist/extension.js
+!extension/dist/extension.js.map
+!extension/dist/web/extension.js
+!extension/dist/web/extension.js.map
```
These could be removed.

mm-broadcom replied:
Fixed in latest commit.

mm-broadcom committed:
Update .vscodeignore

Updated .vscodeignore to remove the unneeded and unused exceptions to the ignore.
